src/main/scala/dsptools/numbers/chisel_types/IntervalTypeClass.scala

```diff
@@ -0,0 +1,278 @@
+  def fromBigDecimal(n: BigDecimal): Interval = fromDouble(n.doubleValue)
+  def fromLong(n: Long): Interval = fromBigInt(BigInt(n))
+  def fromType[B](n: B)(implicit c: ConvertableFrom[B]): Interval = fromDouble(c.toDouble(n))
+  def fromBigInt(n: BigInt): Interval = n.doubleValue.I(0.BP)
+  // If no binary point is specified, use the default one provided by DspContext
+  // TODO: Should you instead be specifying a max width so you can get the most resolution for a given width?
+  def fromDouble(n: Double): Interval = n.I(context.binaryPoint.getOrElse(0).BP)
+  override def fromDouble(d: Double, a: Interval): Interval = {
+    require(a.binaryPoint.known, "Binary point must be known!")
+    d.I(a.binaryPoint)
+  }
+  override def fromDoubleWithFixedWidth(d: Double, a: Interval): Interval = {
+    require(a.binaryPoint.known, "Binary point must be known!")
+    require(a.widthKnown, "Interval width not known!")
+    val sintBits = BigInt(d.toInt).bitLength + 1
+    require(sintBits + a.binaryPoint.get <= a.getWidth, "Lit can't fit in prototype Interval bitwidth")
+    d.I(a.getWidth.W, a.binaryPoint)
+  }
+}
+
+trait ConvertableFromInterval extends ChiselConvertableFrom[Interval] with hasContext {
+  // intPart depends on truncate
+  // asReal depends on shifting fractional bits up
+
+  override def asFixed(a: Interval): FixedPoint = a.asFixedPoint(a.binaryPoint)
+  def asFixed(a: Interval, proto: FixedPoint): FixedPoint = a.asFixedPoint(proto.binaryPoint)
+
+  override def asInterval(a: Interval): Interval = a
+  def asInterval(a: Interval, proto: Interval): Interval = asInterval(a)
+}
+
+trait BinaryRepresentationInterval extends BinaryRepresentation[Interval] with hasContext {
+  def shl(a: Interval, n: Int): Interval = a << n
+  def shl(a: Interval, n: UInt): Interval = a << n
+
+  // Note: This rounds to negative infinity (smallest abs. value for negative #'s is -LSB)
+  def shr(a: Interval, n: Int): Interval = a >> n
+  def shr(a: Interval, n: UInt): Interval = a >> n
+
+  // mul2 consistent with shl
+  // signBit relies on Signed
+
+  // Retains significant digits while dividing
+  override def div2(a: Interval, n: Int): Interval = {
+    require(a.widthKnown, "Interval point width must be known for div2")
+    require(a.binaryPoint.known, "Binary point must be known for div2")
+    val newBP = a.binaryPoint.get + n
+    // Normal shift loses significant digits; this version doesn't
+    val inLong = Wire(Interval((a.getWidth + n).W, newBP.BP))
+    inLong := a
+    val outFull = Wire(Interval(a.getWidth.W, newBP.BP))
+    // Upper n bits don't contain meaningful data following shift, so remove
+    outFull := inLong >> n
+    // Note: The above doesn't rely on tools to expand, shrink correctly; the version below does.
+    // Assumes setBinaryPoint zero-extends. BUT Chisel doesn't seem to get widths properly and
+    // some other ops rely on width correctness... (even though Firrtl is right...)
+    //a.setBinaryPoint(newBP) >> n
```

Comment:
this is unfortunate. do you think our goal should be to ultimately make setBinaryPoint work?